R/driver-snowflake.R

```diff
@@ -87,3 +87,238 @@ setMethod("odbcDataType", "Snowflake",
+  default_paths <- snowflake_default_driver_paths()
+  if (length(default_paths) > 0) {
+    return(default_paths[1])
+  }
+
+  fallbacks <- c("Snowflake", "SnowflakeDSIIDriver")
+  fallbacks <- intersect(fallbacks, odbcListDrivers()$name)
+  if (length(fallbacks) > 0) {
+    return(fallbacks[1])
+  }
+
+  abort(
+    c(
+      "Failed to automatically find Snowflake ODBC driver.",
+      i = "Set `driver` to known driver name or path."
+    ),
+    call = quote(DBI::dbConnect())
+  )
+}
+
+snowflake_default_driver_paths <- function() {
+  if (Sys.info()["sysname"] == "Linux") {
+    paths <- c(
+      "/opt/rstudio-drivers/snowflake/bin/lib/libsnowflakeodbc_sb64.so",
+      "/usr/lib/snowflake/odbc/lib/libSnowflake.so"
+    )
+  } else if (Sys.info()["sysname"] == "Darwin") {
+    paths <- "/opt/snowflake/snowflakeodbc/lib/universal/libSnowflake.dylib"
+  } else {
+    paths <- character()
+  }
+  paths[file.exists(paths)]
+}
+
+snowflake_server <- function(account) {
+  if (nchar(account) == 0) {
+    abort(
+      c(
+        "No Snowflake account ID provided.",
+        i = "Either supply `account` argument or set env var `SNOWFLAKE_ACCOUNT`."
+      ),
+      call = quote(DBI::dbConnect())
+    )
+  }
+  paste0(account, ".snowflakecomputing.com")
+}
+
+snowflake_auth_args <- function(account,
+                                uid = NULL,
+                                pwd = NULL,
+                                authenticator = NULL,
+                                ...) {
+  if (!is.null(uid) && !is.null(pwd)) {
+    return(list(uid = uid, pwd = pwd))
+  } else if (xor(is.null(uid), is.null(pwd))) {
+    abort(
+      c(
+        "Both `uid` and `pwd` must be specified to authenticate.",
+        i = "Or leave both unset to use ambient Snowflake credentials."
+      ),
+      call = quote(DBI::dbConnect())
+    )
+  }
+
+  # Manual override of the authentication method.
+  if (!is.null(authenticator)) {
+    list()
+  }
+
+  # Check for Workbench-provided credentials.
+  sf_home <- Sys.getenv("SNOWFLAKE_HOME")
+  if (grepl("posit-workbench", sf_home, fixed = TRUE)) {
+    token <- workbench_snowflake_token(account, sf_home)
+    if (!is.null(token)) {
+      return(list(authenticator = "oauth", token = token))
+    }
+  }
+
+  # Check for the default token mounted when running in Snowpark Container
+  # Services.
+  if (file.exists("/snowflake/session/token")) {
+    return(list(
+      authenticator = "oauth",
+      token = readLines("/snowflake/session/token", warn = FALSE)
+    ))
+  }
+
+  list()
+}
+
+# Reads Posit Workbench-managed Snowflake credentials from a
+# $SNOWFLAKE_HOME/connections.toml file, as used by the Snowflake Connector for
+# Python implementation. The file will look as follows:
+#
+# [workbench]
+# account = "account-id"
+# token = "token"
+# authenticator = "oauth"
+workbench_snowflake_token <- function(account,
+                                      sf_home = Sys.getenv("SNOWFLAKE_HOME")) {
```

Comment:
```suggestion
workbench_snowflake_token <- function(account, sf_home) {
```

This default made me think that we might need to check its value, but this is actually always set by the calling function

Reply:
Done!